src/highlevel.jl

```diff
@@ -0,0 +1,110 @@
+        write(file, name, data)
+    finally
+        close(file)
+    end
+end
+
+function h5read(filename, name::AbstractString; pv...)
+    local dat
+    fapl = FileAccessProperties(; fclose_degree=:strong)
+    pv = setproperties!(fapl; pv...)
+    file = h5open(filename, "r", fapl)
+    try
+        obj = getindex(file, name; pv...)
+        dat = read(obj)
+        close(obj)
+    finally
+        close(file)
+    end
+    dat
+end
+
+function h5read(filename, name_type_pair::Pair{<:AbstractString,DataType}; pv...)
+    local dat
+    fapl = FileAccessProperties(; fclose_degree=:strong)
+    pv = setproperties!(fapl; pv...)
+    file = h5open(filename, "r", fapl)
+    try
+        obj = getindex(file, name_type_pair[1]; pv...)
+        dat = read(obj, name_type_pair[2])
+        close(obj)
+    finally
+        close(file)
+    end
+    dat
+end
+
+function h5read(
+    filename,
+    name::AbstractString,
+    indices::Tuple{Vararg{Union{AbstractRange{Int},Int,Colon}}};
+    pv...
+)
+    local dat
+    fapl = FileAccessProperties(; fclose_degree=:strong)
+    pv = setproperties!(fapl; pv...)
+    file = h5open(filename, "r", fapl)
+    try
+        dset = getindex(file, name; pv...)
+        dat = dset[indices...]
+        close(dset)
+    finally
+        close(file)
+    end
+    dat
+end
+
+function Base.getindex(parent::Union{File,Group}, path::AbstractString; pv...)
+    haskey(parent, path) || throw(KeyError(path))
+    # Faster than below if defaults are OK
+    isempty(pv) && return open_object(parent, path)
+    obj_type = gettype(parent, path)
+    if obj_type == API.H5I_DATASET
+        dapl = DatasetAccessProperties()
+        dxpl = DatasetTransferProperties()
+        pv = setproperties!(dapl, dxpl; pv...)
+        isempty(pv) || error("invalid keyword options $pv")
+        return open_dataset(parent, path, dapl, dxpl)
+    elseif obj_type == API.H5I_GROUP
+        gapl = GroupAccessProperties(; pv...)
+        return open_group(parent, path, gapl)
+    else#if obj_type == API.H5I_DATATYPE # only remaining choice
+        tapl = DatatypeAccessProperties(; pv...)
+        return open_datatype(parent, path, tapl)
+    end
+end
+
+# Assign syntax: obj[path] = value
+# Create a dataset with properties: obj[path, prop = val, ...] = val
+function Base.setindex!(
+    parent::Union{File,Group}, val, path::Union{AbstractString,Nothing}; pv...
+)
+    need_chunks = any(k in keys(chunked_props) for k in keys(pv))
+    have_chunks = any(k == :chunk for k in keys(pv))
+
+    chunk = need_chunks ? heuristic_chunk(val) : Int[]
+
+    # ignore chunked_props (== compression) for empty datasets (issue #246):
+    discard_chunks = need_chunks && isempty(chunk)
+    if discard_chunks
+        pv = pairs(Base.structdiff((; pv...), chunked_props))
+    else
+        if need_chunks && !have_chunks
+            pv = pairs((; chunk=chunk, pv...))
+        end
+    end
+    write(parent, path, val; pv...)
+end
+
+### Property manipulation ###
```

Comment:
On second thought why not at the end of https://github.com/JuliaIO/HDF5.jl/blob/master/src/properties.jl ? Seems more appropriate especially since that's where the `*Properties` types are defined?

Reply:
I tried that. It does not work. We would need to move the property types up into types.

Reply:
ahh